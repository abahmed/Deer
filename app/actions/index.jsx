// Local libraries
import { ACTION_TYPES } from './types';

/*
*  All notes related actions including payloads, if any
*/

// fetchAllNotes - Retrieve notes from the store

export function fetchAllNotes() {
  return {
    type: ACTION_TYPES.FETCH_ALL_NOTES,
  };
}

// addNote - Add/Save a new note to the store
export function addNote(payload) {
  return {
    type: ACTION_TYPES.ADD_NOTE,
    payload: payload,
  };
}

// const ACTIONS = {
//   addNote: addNote,
//   fetchAllNotes: fetchAllNotes
// }

// export default ACTIONS;

