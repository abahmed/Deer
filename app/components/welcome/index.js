import { connect } from 'react-redux'

import Welcome from './welcome'
import { saveSettings } from '../../actions/settings'

const mapStateToProps = state => ({
  settingsStatus: state.settingsReducer.get('settingsStatus')
})

const mapDispatchToProps = dispatch => ({
  saveSettings: newSettings => dispatch(saveSettings(newSettings))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Welcome)
