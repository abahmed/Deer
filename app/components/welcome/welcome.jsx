import React, { Component } from 'react'
import PropTypes from 'prop-types'

// UI components.
import Typography from '@material-ui/core/Typography'
import Grid from '@material-ui/core/Grid'
import Select from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'
import FormControl from '@material-ui/core/FormControl'
import InputLabel from '@material-ui/core/InputLabel'
import Button from '@material-ui/core/Button'
import Paper from '@material-ui/core/Paper'
import Fade from '@material-ui/core/Fade'
import NavigateNext from '@material-ui/icons/NavigateNext'
import Styles from './style'

// UI wrappers.
import { withStyles } from '@material-ui/core/styles'
import { withTheme } from '@material-ui/core/styles'

import { SUPPORTED_LANGS } from '../../i18n/locales'
import { LANG_LIST } from './langList'
import {
  setNotFirstTimeFlag,
  getDefaultLanguage
} from '../../utils/api.electron'


class Welcome extends Component {
  constructor (props) {
    super()
    this.state = {
      fadeIn: true,
      settings: {
        language: getDefaultLanguage()
      }
    }

    // Away from state to avoid re-rendering when their values change.
    this.langIndex = 0
    // Used to stop after navigating away from this component.
    this.timer = 0

    this.onLanguageChange = this.onLanguageChange.bind(this)
    this.onSaveSettings = this.onSaveSettings.bind(this)
  }

  componentWillMount () {
    this.updateLang()
  }

  componentWillUnmount () {
    // Stopping the timer to call updateLang
    clearTimeout(this.timer)
  }

  // Update fadeIn boolean and index, so UI is re-rendered with next values.
  updateLang() {
    // Update next lang index after fading out to be used for next fade-in.
    if (!this.state.fadeIn)
      this.updateNextLangIndex()

    this.timer = setTimeout(() => {
      this.toggleFade()
      this.updateLang()
    }, this.state.fadeIn ? 3000 : 180)
  }

  // Update the index value to point to the next element in the langList
  // array, and if it has reached the end, it's set to zero.
  updateNextLangIndex() {
    let index = this.langIndex + 1
    if (index >= LANG_LIST.length) {
      index = 0
    }
    this.langIndex = index
  }

  // Used for inverting values of fadeIn
  toggleFade() {
    this.setState(state => {
      return {
        ...state,
        fadeIn: !state.fadeIn
      }
    })
  }

  // Called when user clicks on next button.
  onSaveSettings () {
    // Saving new Language if it's different from defaults.
    if (this.state.settings.language !== getDefaultLanguage()) {
      this.props.saveSettings({language: this.state.settings.language})
    }

    setNotFirstTimeFlag()

    // Navigate to home page.
    this.props.history.push('/')
  }

  // Called when user changes language in select box.
  onLanguageChange (event) {
    this.setState(state => {
      return {
        ...state,
        settings: {
          language: event.target.value
        }
      }
    })
    event.persist()
  }

  render () {
    const lang = LANG_LIST[this.langIndex]
    const { classes } = this.props
    return (
      <Grid container justify='center' className={classes.root}>
        <Paper className={classes.paper}>
            <img
              src={require('../../assets/images/Deer-256.png')}
            />
            <Fade in={this.state.fadeIn}>
              <Typography
                variant='h5'
                component='h5'
                className={classes.item}>
                {lang.welcome}
              </Typography>
            </Fade>
            <FormControl className={classes.formControl}>
              <InputLabel htmlFor='language'>
                {lang.selectLang}
              </InputLabel>
              <Select
                value={this.state.settings.language}
                onChange={this.onLanguageChange}
                name='language'
              >
                {SUPPORTED_LANGS.map((lang, index) => (
                  <MenuItem key={index} value={lang.code}>
                    {lang.language}
                  </MenuItem>
                ))}
              </Select>
            </FormControl>
            <Button
              variant='contained'
              color='primary'
              className={classes.item}
              onClick={this.onSaveSettings}>
            {lang.nextBtn}
            <NavigateNext className={classes.nextIconSize}/>
            </Button>
        </Paper>
      </Grid>
    )
  }
}

Welcome.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired
}
export default withTheme()(withStyles(Styles)(Welcome))