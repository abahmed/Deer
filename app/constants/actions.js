export const ACTIONS = {
  GET_NEXT_LANG: 'SET_NEXT_LANG',
  TOGGLE_FADE: 'TOGGLE_FADE',
  UPDATE_EDITOR_STATE: 'UPDATE_EDITOR_STATE',
  FETCH_ALL_NOTES: 'FETCH_ALL_NOTES',
  UPDATE_NOTE_LIST: 'UPDATE_NOTE_LIST',
  SET_ACTIVE_NOTE: 'SET_ACTIVE_NOTE',
  UPDATE_NEW_NOTE_STATE: 'UPDATE_NEW_NOTE_STATE',
  UPDATE_SAVE_NOTE_STATE: 'UPDATE_SAVE_NOTE_STATE',
  ADD_NOTE: 'ADD_NOTE'
}
