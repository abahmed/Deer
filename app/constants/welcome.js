export const welcome = [
  {
    lang: 'English',
    welcome: 'Welcome',
    nextBtn: 'Next'
  },
  {
    lang: 'Arabic',
    welcome: 'أهلا بك',
    nextBtn: 'التالى'
  },
  {
    lang: 'French',
    welcome: 'Bienvenue',
    nextBtn: 'Prochain'
  },
  {
    lang: 'Spanish',
    welcome: 'Bienvenido',
    nextBtn: 'Siguiente'
  },
  {
    lang: 'Italian',
    welcome: 'Benvenuto',
    nextBtn: 'Il prossimo'
  },
  {
    lang: 'German',
    welcome: 'Willkommen',
    nextBtn: 'Nächster'
  },
  {
    lang: 'Russian',
    welcome: 'Добро пожаловать',
    nextBtn: 'Далее'
  },
  {
    lang: 'Chinese',
    welcome: '欢迎',
    nextBtn: '下一个'
  },
  {
    lang: 'Japanese',
    welcome: 'ようこそ',
    nextBtn: '次'
  },
  {
    lang: 'Hindi',
    welcome: 'स्वागत हे',
    nextBtn: 'आगामी'
  }
]
