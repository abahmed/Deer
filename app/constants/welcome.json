[
  {
    "lang": "English",
    "welcome": "Welcome",
    "selectLang": "Select your language",
    "nextBtn": "Next"
  },
  {
    "lang": "Arabic",
    "welcome": "مرحبا",
    "selectLang": "اختر لغتك",
    "nextBtn": "التالى"
  }
]
