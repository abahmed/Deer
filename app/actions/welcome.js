import { createAction } from 'redux-actions'
import { ACTIONS } from '../constants/actions'

// Used for updating index value of the langList.
export const getNextLang = createAction(ACTIONS.GET_NEXT_LANG)

// Used for inverting values of fadeIn.
export const toggleFade = createAction(ACTIONS.TOGGLE_FADE)

// Setting the timer to a variable so to be able to stop after navigating away from Welcome component.
let timer = 0
// Async action that is fired from time to time update fadeIn boolean and
// index, so UI is rerendered with next values.
export const updateLang = () => (dispatch, getState) => {
  // Only update the index when fading in, as fading out will not be shown.
  const state = getState().welcomeReducer
  let isFadeIn = state.fadeIn
  if (isFadeIn) { dispatch(getNextLang()) }

  timer = setTimeout(() => {
    dispatch(toggleFade())
    dispatch(updateLang())
  }, isFadeIn ? 3000 : 180)
}

// Stopping the timer to call updateLang
export const stopUpdateLang = () => () => {
  clearTimeout(timer)
}
