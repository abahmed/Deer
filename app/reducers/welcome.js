import { ACTIONS } from '../constants/actions'
import { welcome } from '../constants/welcome'

const INITIAL_STATE = {
  index: -1,
  fadeIn: true,
  langList: welcome
}

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case ACTIONS.GET_NEXT_LANG:
      // Update the index value to point to the next element in the langList
      // array, and if it reached to the end, it's set to zero.
      let index = -1
      if (state.langList.length !== 0) {
        index = state.index + 1
        if (index >= state.langList.length) { index = 0 }
      }
      return {
        ...state,
        index: index
      }
    case ACTIONS.TOGGLE_FADE:
      // Invert the value of fadeIn (boolean).
      return {
        ...state,
        fadeIn: !state.fadeIn
      }
    default:
      return state
  }
}
