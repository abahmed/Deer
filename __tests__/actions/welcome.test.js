import configureStore from 'redux-mock-store'

// Actions to be tested
import * as welcomeActions from '../../app/actions/welcome'

const mockStore = configureStore()
const store = mockStore()

describe('welcome_actions', () => {
  beforeEach(() => {
    store.clearActions()
  })

  describe('getNextLang', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(welcomeActions.getNextLang())
      expect(store.getActions()).toMatchSnapshot()
    })
  })

  describe('toggleFade', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(welcomeActions.toggleFade())
      expect(store.getActions()).toMatchSnapshot()
    })
  })
})
