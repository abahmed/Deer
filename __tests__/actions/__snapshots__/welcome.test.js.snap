// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`welcome_actions getNextLang Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "type": "SET_NEXT_LANG",
  },
]
`;

exports[`welcome_actions toggleFade Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "type": "TOGGLE_FADE",
  },
]
`;
